types/decimal.go

```diff
@@ -0,0 +1,356 @@
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }             // greater than
+func (d Dec) GTE(d2 Dec) bool   { return (d.Int).Cmp(d2.Int) >= 0 }             // greater than or equal
+func (d Dec) LT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == -1 }            // less than
+func (d Dec) LTE(d2 Dec) bool   { return (d.Int).Cmp(d2.Int) <= 0 }             // less than or equal
+func (d Dec) Neg() Dec          { return Dec{new(big.Int).Neg(d.Int)} }         // Is equal to zero
+func (d Dec) Add(d2 Dec) Dec    { return Dec{new(big.Int).Add(d.Int, d2.Int)} } // addition
+func (d Dec) Sub(d2 Dec) Dec    { return Dec{new(big.Int).Sub(d.Int, d2.Int)} } // subtraction
+
+// multiplication
+func (d Dec) Mul(d2 Dec) Dec {
+	mul := new(big.Int).Mul(d.Int, d2.Int)
+	chopped := BankerRoundChop(mul, Precision)
+
+	if chopped.BitLen() > 255 {
+		panic("Int overflow")
+	}
+	return Dec{chopped}
+}
+
+// quotient
+func (d Dec) Quo(d2 Dec) Dec {
+
+	// multiply precision twice
+	mul := new(big.Int).Mul(d.Int, precisionReuse)
+	mul.Mul(mul, precisionReuse)
+
+	quo := new(big.Int).Quo(mul, d2.Int)
+	chopped := BankerRoundChop(quo, Precision)
+	return Dec{chopped}
+}
+
+func (d Dec) String() string {
+	str := d.ToLeftPaddedWithDecimals(Precision)
+	placement := len(str) - Precision
+	if placement < 0 {
+		panic("too few decimal digits")
+	}
+	return str[:placement] + "." + str[placement:]
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPaddedWithDecimals(totalDigits int8) string {
+	intStr := d.Int.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPadded(totalDigits int8) string {
+	chopped := BankerRoundChop(d.Int, Precision)
+	intStr := chopped.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+//     ____
+//  __|    |__   "chop 'em
+//       ` \     round!"
+// ___||  ~  _     -bankers
+// |         |      __
+// |       | |   __|__|__
+// |_____:  /   | $$$    |
+//              |________|
+
+// nolint - go-cyclo
+// chop of n digits, and banker round the digits being chopped off
+// Examples:
+//   BankerRoundChop(1005, 1) = 100
+//   BankerRoundChop(1015, 1) = 102
+//   BankerRoundChop(1500, 3) = 2
+func BankerRoundChop(d *big.Int, n int64) (chopped *big.Int) {
+
+	// remove the negative and add it back when returning
+	if d.Sign() == -1 {
+		d = new(big.Int).Neg(d)
+		defer func() {
+			chopped = new(big.Int).Neg(chopped)
+		}()
+	}
+
+	// get the trucated quotient and remainder
+	quo, rem := big.NewInt(0), big.NewInt(0)
+	quo, rem = quo.QuoRem(d, precisionReuse, rem)
+
+	if rem.Sign() == 0 { // remainder is zero
+		return quo
+	}
+
+	lenWhole := len(d.String())
```

Comment:
Do we really need to convert to a string here? Shouldn't the length be calculable - `Precision` (characters after decimal), plus one for `.` (decimal), plus `floor (log_10 d)` (characters before the decimal), or something like that?

Reply:
did a bunch of research there isn't an obvious way of doing this... the only thing I could see was converting to a float64 which is far less than ideal. doesn't look like there are log functions for big.Int (or big.Float!) but either way I think this conversion is fine and efficient - remember this is simply a big.Int.String() not a dec.String() which means it doesn't have a decimal point or any of the fun logic in dec.String().

I'm going to leave this

Reply:
Im confused why we need strings. Why is the result different from:
```
BankerStatus := rem.Cmp(5*10^9)
If BankerStatus == -1 {
 Return quo 
} if BankerStatus == 0 {
 Parity check 
} else 
{ 
Return quo + 1
}
```
No strings necessary, right? (Sorry for bad syntax, on mobile)

Reply:
yes!  ----- yeah okay so there was a subtle change that was made (and probs should have been communicated better, but more importantly caught by a linter) which was to modify this function to instead chopping off `n` amount of digits from the end - I actually fully agree with this change as a good point of optimization because we actually ONLY ever round by the precision. -> what the linter should have caught was the fact that `n` was still included as a function input even though it was not used anywhere -> if we do have a linter for this I suspect it wasn't caught because we DO have one recursive function call in here, so even though the variable isn't used for any calculations, it's actually used to pass on to the next recursion (sneaky!). But yeah this function probably should have been fully refactored with that in mind when it was updated to only ever chop a `Precision` number of digits ---------------- BUT you're totally right, it really don't need a string or any of the "fun" logic to determine the number of leading zeros in the remainder. Thanks! updating now. (👏  woot less complexity )